src/terminal-window.c

```diff
@@ -3058,7 +3058,29 @@ notebook_button_press_cb (GtkWidget *widget,
     /* If the event is a double click, display the set title dialog */
     if (event->type == GDK_DOUBLE_BUTTON_PRESS)
     {
-        terminal_set_title_callback (NULL, window);
+        /*No terminal window should ever be smaller than this (2 rows min in Adwaita & usually 4-5)
+         *No tab header should ever be larger than this even in Adwaita (which has high tab headers)
+         *We need this to distinguish a tab header's window from the main page window of the tab
+         */
+        const int max_tab_bar_height = 40;
+
+        GdkWindow *tabwin;
+        GdkSeat *seat;
+        GdkDevice *pointer;
+        gint height = 0;
+        GdkDisplay *display = gdk_screen_get_display (gdk_screen_get_default());
+
+        seat = gdk_display_get_default_seat (display);
+        pointer = gdk_seat_get_pointer (seat);
+
+        tabwin = gdk_device_get_window_at_position (pointer, NULL, NULL);
+        gdk_window_get_geometry (tabwin, NULL, NULL, NULL, &height);
+        /*If the window under the pointer isn't tall enough to be
+         *the terminal window or main tab page window yet is in the notebook,
+         *it is the tab header, the only thing we want to doubleclick on to set the title
+         */
+        if (height < max_tab_bar_height)
+            terminal_set_title_callback (NULL, window);
```

Comment:
I am afraid this makes way too many assumptions in what a tab might be or not be like, as well as requiring the tab having an actual window (IIUC), which could potentially change in the future.

What about doing it like that:
```diff
diff --git a/src/terminal-window.c b/src/terminal-window.c
index 6517d8d..e86172f 100644
--- a/src/terminal-window.c
+++ b/src/terminal-window.c
@@ -551,10 +551,12 @@ find_tab_num_at_pos (GtkNotebook *notebook,
         max_x = x_root + tab_allocation.x + tab_allocation.width;
         max_y = y_root + tab_allocation.y + tab_allocation.height;
 
-        if ((tab_pos == GTK_POS_TOP || tab_pos == GTK_POS_BOTTOM) && screen_x <= max_x)
+        if ((tab_pos == GTK_POS_TOP || tab_pos == GTK_POS_BOTTOM) && screen_x <= max_x &&
+            screen_y >= y_root + tab_allocation.y && screen_y <= max_y)
             return page_num;
 
-        if ((tab_pos == GTK_POS_LEFT || tab_pos == GTK_POS_RIGHT) && screen_y <= max_y)
+        if ((tab_pos == GTK_POS_LEFT || tab_pos == GTK_POS_RIGHT) && screen_y <= max_y &&
+            screen_x >= x_root + tab_allocation.x && screen_x <= max_x)
             return page_num;
 
         page_num++;
@@ -3056,7 +3058,8 @@ notebook_button_press_cb (GtkWidget *widget,
     }
 
     /* If the event is a double click, display the set title dialog */
-    if (event->type == GDK_DOUBLE_BUTTON_PRESS)
+    if (event->type == GDK_DOUBLE_BUTTON_PRESS &&
+        find_tab_num_at_pos (notebook, (int) event->x_root, (int) event->y_root) >= 0)
     {
         terminal_set_title_callback (NULL, window);
 
```

We already have a function that does almost this, and improving it to validate the tab area seems easy enough.
Trying it quickly seems to work fine for me, but please tell me if there's a technical reason you didn't do it like that?

Reply:
`gtk_notebook_get_tab_pos ()` only tells which edge the tab headers are drawn on, NOT whether the pointer is in the header or not. 

I do see issues with having to resort to this sort of coding, but I cannot find a function in GtkNoteBook to tell whether the pointer is inside the header or not. A screwup here will either disable the popup entirely or bring back the unwanted behavior of opening the tab renaming dialog from clicking on the tab.

Basically, in GtkNoteBook, it seems the tab or page so far as interaction goes means the WHOLE tab or page. You can click on any showing portion and no distinction between header and rest of page is made.

I am also entirely OK with reverting https://github.com/mate-desktop/mate-terminal/commit/b9e68e715f95aac477a1fe7e189a5121091a1c68 which created https://github.com/mate-desktop/mate-terminal/issues/444 in the first place. Note however that I myself neither use nor rename tabs in terminal so I am not the best to make the UI decision on this.

If someone has a better way to do this that works, I welcome that as well. I don't think the currenrt behavior is good for a major release UNLESS there is no in-terminal application in use anywhere that uses a double-click inside the terminal for anything. In that event https://github.com/mate-desktop/mate-terminal/issues/444 may be of no importance

Reply:
I don't really understand what you are afraid of with the code I suggest: `get_tab_pos()` is only used to find out whether to check on X or Y direction; the actual positions use the **tab's** widget allocation to make sure it's in there. Admittedly it looks incorrect for pos=BOTTOM and pos=RIGHT, but AFAICT we don't have a setting for tab position, so it's always gonna be at the top -- and it's easy to fix.

Would you be happier with something like this?
```diff
diff --git a/src/terminal-window.c b/src/terminal-window.c
index 6517d8d..bcae49d 100644
--- a/src/terminal-window.c
+++ b/src/terminal-window.c
@@ -519,27 +519,23 @@ escape_underscores (const char *name)
 }
 
 static int
-find_tab_num_at_pos (GtkNotebook *notebook,
+find_tab_num_at_pos (GtkNotebook *nb,
                      int screen_x,
                      int screen_y)
 {
-    GtkPositionType tab_pos;
     int page_num = 0;
-    GtkNotebook *nb = GTK_NOTEBOOK (notebook);
     GtkWidget *page;
     GtkAllocation tab_allocation;
 
-    tab_pos = gtk_notebook_get_tab_pos (GTK_NOTEBOOK (notebook));
-
     while ((page = gtk_notebook_get_nth_page (nb, page_num)))
     {
         GtkWidget *tab;
-        int max_x, max_y, x_root, y_root;
+        int x_root, y_root;
 
         tab = gtk_notebook_get_tab_label (nb, page);
         g_return_val_if_fail (tab != NULL, -1);
 
-        if (!gtk_widget_get_mapped (GTK_WIDGET (tab)))
+        if (!gtk_widget_get_mapped (tab))
         {
             page_num++;
             continue;
@@ -548,13 +544,10 @@ find_tab_num_at_pos (GtkNotebook *notebook,
         gdk_window_get_origin (gtk_widget_get_window (tab), &x_root, &y_root);
 
         gtk_widget_get_allocation (tab, &tab_allocation);
-        max_x = x_root + tab_allocation.x + tab_allocation.width;
-        max_y = y_root + tab_allocation.y + tab_allocation.height;
-
-        if ((tab_pos == GTK_POS_TOP || tab_pos == GTK_POS_BOTTOM) && screen_x <= max_x)
-            return page_num;
-
-        if ((tab_pos == GTK_POS_LEFT || tab_pos == GTK_POS_RIGHT) && screen_y <= max_y)
+        if (screen_x >= x_root + tab_allocation.x &&
+            screen_x <= x_root + tab_allocation.x + tab_allocation.width &&
+            screen_y >= y_root + tab_allocation.y &&
+            screen_y <= y_root + tab_allocation.y + tab_allocation.height)
             return page_num;
 
         page_num++;
@@ -3056,7 +3049,8 @@ notebook_button_press_cb (GtkWidget *widget,
     }
 
     /* If the event is a double click, display the set title dialog */
-    if (event->type == GDK_DOUBLE_BUTTON_PRESS)
+    if (event->type == GDK_DOUBLE_BUTTON_PRESS &&
+        find_tab_num_at_pos (notebook, (int) event->x_root, (int) event->y_root) >= 0)
     {
         terminal_set_title_callback (NULL, window);
 
```

Another solution could be having the mouse click handling on the tab widget itself rather than the notebook itself.  This would require one handler per tab, but all "where did you click" handling could be removed -- yet you'd still have to find out what tab it is related to, bu that's easy enough.

Anyway, I don't really care myself for the feature itself, but somebody wanted it enough to implement it, so it might be of importance to them.

Reply:
Just tested your code, it seems to work. Basically I wasn't able from the documentation for GtkNotebook to figure out how to get the tab allocation, but you've got something that works. Will entirely replace my previously proposed fix with this. THANKS!